### ISSUE TITLE
mac homebrew support

### ISSUE THREAD

**Issue by bingalls** [dvehrs/podget#25]
0.8.4 requires the Gnu fork of sed. For BSD sed users, such as Mac OSX, add the following near the top of podget, and replace every instance of *sed* with *${SED}* .
I don't currently have *Bsd to test with, but these also have gnu sed packages. Additional homebrew packages, such as wget may be required, but your code comment covers these.
Finally, you must modify Makefile to install to /usr/local/bin/podget due to security restrictions on /usr/bin for Mac OSX 10.11.0 (El Capitain) or newer.

```bash
SED=`which sed`
if [ 'Darwin' = `uname` ];then
  SED=`which gsed`
  if [ ! ${SED} ];then
    echo 'Gnu sed required. Try `brew install gnu-sed` #See https://brew.sh/ for details'
    exit
  fi
fi
```

Note that podget tests as working on my Mac, after using gnu sed.

**Comment by dvehrs** [comment]
OK, interesting point and should be a relatively easy fix.

Your example looks good on the surface but there are a few things I would do different.

1.  Subshells:  I hate using back-ticks for subshells.  They can be hard to distinguish depending on what font the user may choose.  Therefore I tend to prefer $(...) over `...` for subshells.

2, Which:  Which is not a builtin and doesn't always exit with proper exit codes depending on the platform you use it on.   I've migrated away from using which to using type or hash for bash.  Simple change and helps simplify the script and it might reduce the number of subshells we need to use.

Good concise explanation of why you should use hash, type or command instead of which:
https://stackoverflow.com/questions/592620/check-if-a-program-exists-from-a-bash-script

Now I haven't tested this but I'm thinking maybe something like:

```
if [ "$(uname)" = "Darwin" ]; then
  if ! hash gsed >/dev/null 2>&1; then
    echo "GNU Sed Required."
    echo "Try 'brew install gnu-sed'  (see htts://brew.sh for details)"
    exit 1
  else
    SED="gsed"
  fi
else
  SED="sed"
fi
```

Now I haven't tested this but it does reduce subshells and respects the PATH variable (as which would do).  

The Makefile is an interesting issue.  I haven't spent that much time with the Makefile but we may need to find a way to detect that we're running on Darwin and change the install location.  Much like you did with the uname command.  

Definitely some things to think about.  I haven't forked a development branch just for Mac OSX experiments but I will soon.   Then we can start playing with some of these ideas.

Thanks!

**Comment by dvehrs** [comment]
OK, started a new branch "dev-machomebrew" to start experimenting with these issues.  No new code yet but we've got a place to play....

**Comment by bingalls** [comment]
I tested your code into pull request https://github.com/dvehrs/podget/pull/27 which should close this issue.

**Comment by dvehrs** [comment]
Closes the issue for Mac but creates an issue for everyone else.  

We cannot change the Makefile in a way that works ONLY for Mac but doesn't respect other platforms standards.  As I mentioned above, we need to find a way to have the Makefile recognize different prefix settings when it is installed on different platforms.

Second,  any bug report or pull request should deal with ONE ISSUE.   If it deals with more than one, the odds of good ideas being refused because of one minor issue go up.   So because your pull request has fix that isn't portable....

**Comment by dvehrs** [comment]
Second issue with your pull request.

You use $SED as your variable.  However, I have standardized on wrapping variables in ${...}.  So to adhere to the standard I use, it should be ${SED}.    

Third issue:  Should the modification of MOST_RECENT on line 286 really be included in this pull request or not?  Was it just a personalization that wasn't meant for everyone?

**Comment by dvehrs** [comment]
OK I just added a commit to the dev-machomebrew branch with a new modified Makefile that uses "uname -s" to detect the OS and then set the prefix value accordingly.  After quick tests on Linux, it works as expected here.   Can you download the Makefile from that branch of test it on Mac OS?

Thanks

**Comment by dvehrs** [comment]
OK, I won't pretend that I knew the method for OS detection that I used in the Makefile before tonight.  I found the answer we needed on Stackoverflow.

https://stackoverflow.com/questions/714100/os-detecting-makefile

**Comment by bingalls** [comment]
I updated and tested the results of your code review.
You might want to reconsider the Recent Files setting, as naive use sucks up bandwidth & time.

**Comment by dvehrs** [comment]
The script has a default download all items because any limit that we choose isn't likely to be shared by all users.   This default can be changed by either modifying MOST_RECENT in the users podgetrc file or by specifying a limit on the command line (-r or --recent options).     So changing it within the script isn't really necessary in my opinion.

**Comment by bingalls** [comment]
I pulled your changes down into mine. If you don't want to limit MOST_RECENT, you should likely add instructions / reminder in your README.md, as this can give a poor user experience for naive users

**Comment by dvehrs** [comment]
The issue of whether or not MOST_RECENT should have a different default is in no way related to the core topic of this thread.  So it is completely wrong to bring it up here.   Others may have a valid opinion on this issue and never look in this thread because this is a thread about Mac OS and they don't run a Mac.     Simple rule for submitting issues, one topic per thread.   That makes them easy to discuss and close when resolved.   Threads with many topics can get closed with some unresolved by accident.  We aren't going to come anywhere near Githubs maximum number of threads, so if you care about this issue start a new thread.

**Comment by dvehrs** [comment]
Two issues with the pull request.

1.  You haven't answered if the removal of CTRL-M from line 1145 (original position, or 1163 with your changes) was intentional or accidental.   If it was intentional, have you tested converting files from DOS to UNIX formats?  

2.  Whether or not MOST_RECENT needs a more explicit warning is in no way related to the core focus of this patch.   Including it just confuses things.   This branch should be focused on fixing things that are Mac OS specific.  Even then we should be careful about trying to do too much.  We've already looked into the Makefile issue and we're close to having a proposed solution to the SED issue.  Lets not make this thread any more complicated than it needs to be.

**Comment by bingalls** [comment]
Sorry that I missed the Ctrl-M issue.
It was not intentional.
I have not tested dos / Unix conversion.
I've tested essential pod catching, only.
Do you have a test plan?
I saw no unit tests.

**Comment by dvehrs** [comment]
OK.  There is no formal test plan or unit tests. I've never determined a good way to do them.  Mainly for release testing I've developed a list of feeds with unique issues that I run against to test it.  Some of the issues like the Mac OS related issues are difficult for me to test because I do not have a Mac OS system of my own to test on.  So for those issues, I am dependent upon the cooperation of users who submit issues to help me test potential fixes.  Much as we've done here.  Unfortunately, DOS/Unix conversions are commonly in that category as well because I no longer have a personal Windows system to test them on.  So I'm hesitant to change that code without some clear evidence that there is an issue.  Now, I haven't received any requests for ASX playlist issues or support in years so it is entirely possible that the best thing to do is completely remove that function.  On the other hand, it's not causing any harm sitting in an isolated function so it stays.  

I do run the script through shellcheck prior to every release to help catch any syntax, semantic or other issues.  Now most of those I've fixed but if you do it, you will see that there are three outstanding issues focused on replacing SED usage with ${variable/search/replace} instead.  Unfortunately in those three cases, I have not worked out good ways to do that... yet.  

Now would you like to fix those two issues before I merge your pull request or merge it as is and then fix them afterwards?

**Comment by dvehrs** [comment]
Wait I just notice a third issue on line 705 of your changes.

$${SED} should be ${SED}

**Comment by dvehrs** [comment]
Thanks for the changes.  I've accepted the pull request into dev-machomebrew.  So its ready for testing.

I have one minor stylistic question.  On line 638, you say that the Portability code is so that it can run on "BSD/Mac OSX", but I was under the impression that Homebrew was a solution for Mac OS only.   Does it work with BSDs too?  And if so do we need to expand the uname matching to include them?  Or would it be simpler to remove the BSD from the comment, figure out what it needs for uname matching then add it back later?

Also I wonder if we shouldn't add a small segment to the INSTALL doc that explains that Mac users will need GNU Sed and explain how to get it there rather than waiting for the first time the user runs Podget.   I've made an initial modification to the INSTALL file, what do you think?  Too wordy?  Not wordy enough?

**Comment by dvehrs** [comment]
Also should we add a general dependencies section to the INSTALL doc so that users will know what is required?

**Comment by dvehrs** [comment]
And am I violating my own rule of not mixing up new topics in threads here.....   probably guilty.

**Comment by dvehrs** [comment]
OK, initial testing with about 9 feeds on Linux looks good.

**Comment by bingalls** [comment]
The INSTALL doc looks good, in general.

However: it is missing the part about modifying the destination 
directory to /usr/local in the Makefile.

Also, the bottom of the INSTALL doc talks about the installed package, 
which should likely move to the Debian section of the document.

Note that uncommenting the catget section, lines 1024 - 1031 still works.

However, it outputs many of the following warnings, which looks annoying:


expr: invalid character range
/usr/local/bin/podget: line 1025: ((: > 0 : syntax error: operand 
expected (error token is "> 0 ")

**Comment by dvehrs** [comment]
For the INSTALL doc:
1.  Part about modifying the Makefile should be completely unneeded now that we have a makefile that automatically changes that based on the OS detected.
2.  The piece at the bottom applies to all installs so I moved it up to the Source Install section.

As for the Catradio section:
1.  I do not get the errors you see.  Is this another version issue similar to what was seen with sed vs GNU sed?   if I run 'expr --version' I get 'expr (GNU coreutils) 8.2.6'.   What do you see?
2.  Further testing with recent Catradio podcast URLs seems to reveal that they have changed their URL structure to include a bit var19=24/04/17.  This little date segment is using the '/' character in a way that is breaking the filename extraction earlier.  I haven't been able to fix FILENAME_FORMATFIX3 yet and I'm not entirely sure if it can be done with the current podget implementation.   So the best bet may be to disable this check completely.   If you look in your $HOME/.podget/podgetrc file for the line FILENAME_FORMATFIX3=1, just change that to a 0 and it won't attempt to use it.    

For the portability section:
1.  How do you feel about changing 'BSD/Mac OSX' to read just 'Mac OSX' for the reasons I outlined above?

**Comment by dvehrs** [comment]
Also I changed the default value for FILENAME_FORMATFIX3=0.  So it should not be enabled for any new podgetrc files that are generated.

**Comment by dvehrs** [comment]
Just found an interesting webpage on installing GNU tools on Mac OS, and it is possible we need to detect/install a few more packages.

https://www.topbug.net/blog/2013/04/14/install-and-use-gnu-command-line-tools-in-mac-os-x/

Now I develop and test Podget on Debian.   The following dependencies may need to be installed from brew for Mac OS:
1.  Coreutils 
2. Findutils
3. grep
4. sed
5. wget

Now, it appears from your testing that we do not need all of them.  That the versions included with Mac OS are frequently close enough.   It is possible that all we will need to warn the Mac OS users to install via brew are gnu-sed and coreutils.   

And we do not use  the '--with-default-names' for brew then the files should be installed with a 'g' prefix so they should not cause any conflicts on the system.

**Comment by dvehrs** [comment]
Before you attempt another pull request or further testing, you may want to make sure you get a fresh pull from dev-machomebrew.  I am always a little paranoid about branches diverging and so will perform merges down to keep them more or less in sync.  I've merged the changes I've made to the main 'dev' branch down onto the 'dev-machomebrew' branch.   Nothing major but just keeps me confident that the merge back into 'dev' when we are done will happen with as few conflicts that have to be manually accepted as possible.  This time I had a couple conflicts I had to manually massage in that were effective the same change with different commits in different branches.  Ooops.

**Comment by bingalls** [comment]
* Mac OSX is an acceptable label. I mention _BSD_ because you ran into portability issues, by using gnu extensions, particularly in sed.
* The new Makefile works on Mac OSX.
* BSD `expr --version` just returns _--version_ . Flags are a gnu extension
* FILENAME_FORMATFIX3=0 removes the catradio warnings
Only these brew packages are required: No flags are used with brew (install)
* coreutils
* sed
* wget
I've git-pulled down your updates.

**Comment by dvehrs** [comment]
OK, sorry for the delay.

I've dabbled with adding a check for GNU Expr very similar to what we did for Sed.   I've uploaded a new version that appears to work fine for my limited testing here.   Can you give it a run on your Mac?  

As for wget, I'm not sure that the Mac version will not work for us because we don't use many options and those we do use aren't generating any issues yet.

**Comment by dvehrs** [comment]
I also moved the DARWIN checks below the function definitions so we can use the CLEANUP_AND_EXIT function to be sure we delete any session or lock files that might be created.  We're mostly likely doing these checks before we need the cleanup function but who knows about the future....

**Comment by bingalls** [comment]
Updated; tests cleanly!

**Comment by dvehrs** [comment]
Reopened this thread because we're almost done but not quite.

OK, for the CatRadio feeds that precipitated this issue.  I've tested about 6 of their feeds and it appears that the FILENAME_FORMATFIX3 is no longer the best way to fix them.  It appears now that using OPT_FILENAME_LOCATION or OPT_CONTENT_DISPOSITION flags in your serverlist for these feeds does everything that FILENAME_FORMATFIX3 used to.  Additionally, using OPT_FILENAME_RENAME_MDATE adds a prefix to each downloaded file with the modification date of the file.  This makes it easy to list the files.

I've added a note to the podgetrc that is initially installed to let the users know that they do not need to enable FILENAME_FORMATFIX3 and of the other ways to fix these filenames.

**Comment by dvehrs** [comment]
I've pushed my changes to the dev-machomebrew branch.  What do you think?

**Comment by bingalls** [comment]
tests ok

**Comment by dvehrs** [comment]
OK, I merged the changes from this branch to the dev branch.   So assuming everything went well, it's now time to start any testing from it.  

Since any new issues with need a new thread to track them, I think we can now close this one.  

Thanks for the help and input Bruce, we're one step closer to having these changes in a release version.
